packages/backend/src/peripherals/database/ForcedWithdrawRepository.ts

```diff
@@ -0,0 +1,368 @@
+  constructor(database: Database, logger: Logger) {
+    super(database, logger)
+
+    /* eslint-disable @typescript-eslint/unbound-method */
+
+    this.addSent = this.wrapAdd(this.addSent)
+    this.addMined = this.wrapAdd(this.addMined)
+    this.addReverted = this.wrapAdd(this.addReverted)
+    this.addForgotten = this.wrapAdd(this.addForgotten)
+    this.addIncluded = this.wrapAdd(this.addIncluded)
+    this.findByTransactionHash = this.wrapFind(this.findByTransactionHash)
+    this.getByPositionId = this.wrapGet(this.getByPositionId)
+    this.getByStarkKey = this.wrapGet(this.getByStarkKey)
+    this.getByStateUpdateId = this.wrapGet(this.getByStateUpdateId)
+    this.getMinedNotIncluded = this.wrapGet(this.getMinedNotIncluded)
+    this.getJustSentHashes = this.wrapGet(this.getJustSentHashes)
+    this.deleteAll = this.wrapDelete(this.deleteAll)
+    this.deleteAfter = this.wrapDelete(this.deleteAfter)
+
+    /* eslint-enable @typescript-eslint/unbound-method */
+  }
+
+  async addSent(
+    record: ForcedWithdrawTransactionRecord & { timestamp: Timestamp }
+  ): Promise<Hash256> {
+    const knex = await this.knex()
+    await knex.transaction(async (trx) => {
+      await trx('forced_withdraw_transactions')
+        .insert(toRow(record))
+        .onConflict('hash')
+        .ignore()
+      await trx('forced_withdraw_statuses').insert({
+        hash: record.hash.toString(),
+        status: 'sent',
+        timestamp: BigInt(record.timestamp.toString()),
+      })
+    })
+    return record.hash
+  }
+
+  async addMined(
+    record: ForcedWithdrawTransactionRecord & {
+      timestamp: Timestamp
+      blockNumber: number
+    }
+  ): Promise<Hash256> {
+    const knex = await this.knex()
+    await knex.transaction(async (trx) => {
+      await trx('forced_withdraw_transactions')
+        .insert(toRow(record))
+        .onConflict('hash')
+        .ignore()
+      await trx('forced_withdraw_statuses').insert({
+        hash: record.hash.toString(),
+        status: 'mined',
+        timestamp: BigInt(record.timestamp.toString()),
+        block_number: record.blockNumber,
+      })
+    })
+    return record.hash
+  }
+
+  async addReverted(status: {
+    hash: Hash256
+    timestamp: Timestamp
+    blockNumber: number
+  }): Promise<Hash256> {
+    const knex = await this.knex()
+    await knex('forced_withdraw_statuses').insert({
+      hash: status.hash.toString(),
+      status: 'reverted',
+      timestamp: BigInt(status.timestamp.toString()),
+      block_number: status.blockNumber,
+    })
+    return status.hash
+  }
+
+  async addForgotten(status: {
+    hash: Hash256
+    timestamp: Timestamp
+  }): Promise<Hash256> {
+    const knex = await this.knex()
+    await knex('forced_withdraw_statuses').insert({
+      hash: status.hash.toString(),
+      status: 'forgotten',
+      timestamp: BigInt(status.timestamp.toString()),
+    })
+    return status.hash
+  }
```

Comment:
What are forgotten transactions?

Reply:
Sometimes if you send a transaction with a low fee and it doesn't get included in a long time or if you send a transaction that becomes replaced by another transaction (this can happen if you send two transactions with the same nonce and is usually used to increase the tx fee or "cancel" transactions) the nodes drop your transaction and it never ever becomes mined. We have this status to be able to account for this possibility